### DISCUSSION TITLE
Use case for admin read-only attributes in declarative user profile?

### DISCUSSION THREAD

**Discussion by tsaarni** [keycloak/keycloak#25073] [Q&A]
When configuring an attribute as read-only for administrator in declarative user profile, it seems to block setting the attribute also when _initially_ creating the user, not only _editing_ the attribute in an existing user.  In practice, it looks like read-only for admin means that attribute can't exist at all. 

I wonder what is then the scenario when this setting is useful?

**Comment by sample** [comment]
in my case this is attributes from AD, which should to be readonly, but i need it in userprofile (ex. userPrincipalName)

**Comment by ssilvert** [comment]
You can always create the attribute using the REST API.  Note that User Profile is just an overlay that tells the UI what to do.  

Soon we will be releasing an upgrade path for User Profile.   If all goes as planned, you will be able to use the Attributes tab when User Profile is enabled instead of having it always hidden.

So if you have an existing attribute that was created on the attributes page, you could declare that attribute in User Profile and it would now show up only in User Details with its value read-only.  That's a kind of weird use case, but it's a valid one.

I think your point about having the ability to set a read-only value at user creation sounds like a really good use case.  Please create an enhancement request for that one.

**Reply by pedroigor** [reply]
Yeah .... Today, attributes marked as read-only can not be managed at all by a specific role: `user` and/or `admin`.

What is the use case for allowing writes to an attribute only during registration when permissions are set to read-only? Wouldn't be the same as making them writable to admins?

**Reply by tsaarni** [reply]
@ssilvert wrote
> You can always create the attribute using the REST API. Note that User Profile is just an overlay that tells the UI what to do.

I was just about to say that it works differently for me, but then I noticed very recent PR https://github.com/keycloak/keycloak/pull/24937 from @pedroigor and now I can again set attributes even if they are not listed in user profile. Nice! :smile:  

The default for the new "Unmanaged attributes" setting seems `disabled` which could cause backwards compatibility issues after the declarative user profile feature is enabled by default - assume people create new realm with defaults, not knowing it will cause "unmanaged" attributes to be rejected. 

I wonder, will the attributes be rejected also when set via Java API, not via REST API? For example, lets say there is custom provider like this https://github.com/lukaszbudnik/keycloak-ip-authenticator which sets user attributes to control conditional OTP form ([related video](https://youtu.be/u36QK9oyrtM?t=867)). Would it stop working silently, after upgrading and creating a new realm that rejects unmanaged attributes by default?

@pedroigor wrote
> What is the use case for allowing writes to an attribute only during registration when permissions are set to read-only? Wouldn't be the same as making them writable to admins?

I might not have very good example to give right now. On the other hand, I could no figure out what would be the use case for read-only attribute, when there is no way for the attribute to get any value. 

The use case could be something similar to the user creation date, it is set only once. Though that specific data exists already outside of attributes. 

@sample brought up attributes mapped from AD as an answer to my original question, but those are already subject to mappers which are somewhat similar way to control attributes, even to the extent of setting them read-only. It is not clear to me how declarative user profile interacts with user attribute mappers?

**Comment by antikalk** [comment]
One additional thing we would appreciate is that you could somehow configure **custom dynamic checks whether a property is readonly for the user** or not. This would especially be relevant when thinking about this other discussion: https://github.com/keycloak/keycloak/discussions/25341

There could be some user properties that are e.g. assured by an external service and thus should not be editable by the user. But for an alternative way the user could enter the data himself, giving them a lower assurance level, but in this case the user should still be able to edit the values.
